module/support/optimisation/NSGA2Evaluator/src/tasks/WalkEvaluator.cpp

```diff
@@ -0,0 +1,241 @@
+
+    std::unique_ptr<NSGA2FitnessScores> WalkEvaluator::calculate_fitness_scores(bool early_termination,
+                                                                                int generation,
+                                                                                int individual) {
+        double trial_duration =
+            std::chrono::duration_cast<std::chrono::milliseconds>(NUClear::clock::now() - trial_start_time).count();
+
+        auto scores      = calculate_scores();
+        auto constraints = early_termination ? calculate_constraints(trial_duration) : constraints_not_violated();
+
+        NUClear::log<NUClear::DEBUG>("Trial ran for", trial_duration);
+        NUClear::log<NUClear::DEBUG>("SendFitnessScores for generation", generation, "individual", individual);
+        NUClear::log<NUClear::DEBUG>("    scores:", scores[0], scores[1]);
+        NUClear::log<NUClear::DEBUG>("    constraints:", constraints[0], constraints[1]);
+
+        // Create the fitness scores message based on the given results and emit it back to the Optimiser
+        std::unique_ptr<NSGA2FitnessScores> fitness_scores = std::make_unique<NSGA2FitnessScores>();
+        fitness_scores->id                                 = individual;
+        fitness_scores->generation                         = generation;
+        fitness_scores->obj_score                          = scores;
+        fitness_scores->constraints                        = constraints;
+        return fitness_scores;
+    }
+
+    std::vector<double> WalkEvaluator::calculate_scores() {
+        auto robot_distance_travelled = std::fabs(initial_robot_position.x() - robot_position.x());
+        NUClear::log<NUClear::DEBUG>("Distance travelled", robot_distance_travelled);
+        NUClear::log<NUClear::DEBUG>("Max field plane sway", max_field_plane_sway);
+        return {
+            max_field_plane_sway,           // For now, we want to reduce this
+            1.0 / robot_distance_travelled  // 1/x since the NSGA2 optimiser is a minimiser
+        };
+    }
+
+    std::vector<double> WalkEvaluator::calculate_constraints(double trial_duration) {
+        // Convert trial duration limit to ms, add 1 for overhead
+        const auto overhead = std::chrono::seconds(1);
+        double max_trial_duration =
+            (std::chrono::duration_cast<std::chrono::milliseconds>(trial_duration_limit + overhead)).count();
+
+        return {
+            trial_duration - max_trial_duration,  // Punish for falling over, based on how long the trial took
+                                                  // (more negative is worse)
+            0.0                                   // Second constraint unused, fixed to 0
+        };
+    }
+
+    std::vector<double> WalkEvaluator::constraints_not_violated() {
+        return {
+            0.0,  // Robot didn't fall
+            0.0   // Second constraint unused, fixed to 0
+        };
+    }
+
+}  // namespace module::support::optimisation
```

Comment:
Same comment - walk, strafe and rotation have very similar code structure so we might be able to think of a smart way to condense it all